commit 3c733c78d6f8e50296d6e248968b6516c92a53ca
Author: Natalia <124304+nessita@users.noreply.github.com>
Date:   Mon Aug 19 14:47:38 2024 -0300

    [5.1.x] Fixed CVE-2024-45231 -- Avoided server error on password reset when email sending fails.
    
    On successful submission of a password reset request, an email is sent
    to the accounts known to the system. If sending this email fails (due to
    email backend misconfiguration, service provider outage, network issues,
    etc.), an attacker might exploit this by detecting which password reset
    requests succeed and which ones generate a 500 error response.
    
    Thanks to Thibaut Spriet for the report, and to Mariusz Felisiak, Adam
    Johnson, and Sarah Boyce for the reviews.

diff --git a/django/contrib/auth/forms.py b/django/contrib/auth/forms.py
index a11668944a..edf672a6e5 100644
--- a/django/contrib/auth/forms.py
+++ b/django/contrib/auth/forms.py
@@ -1,3 +1,4 @@
+import logging
 import unicodedata
 
 from django import forms
@@ -16,6 +17,7 @@ from django.utils.translation import gettext
 from django.utils.translation import gettext_lazy as _
 
 UserModel = get_user_model()
+logger = logging.getLogger("django.contrib.auth")
 
 
 def _unicode_ci_compare(s1, s2):
@@ -392,96 +394,101 @@ class AuthenticationForm(forms.Form):
 class PasswordResetForm(forms.Form):
     email = forms.EmailField(
         label=_("Email"),
         max_length=254,
         widget=forms.EmailInput(attrs={"autocomplete": "email"}),
     )
 
     def send_mail(
         self,
         subject_template_name,
         email_template_name,
         context,
         from_email,
         to_email,
         html_email_template_name=None,
     ):
         """
         Send a django.core.mail.EmailMultiAlternatives to `to_email`.
         """
         subject = loader.render_to_string(subject_template_name, context)
         # Email subject *must not* contain newlines
         subject = "".join(subject.splitlines())
         body = loader.render_to_string(email_template_name, context)
 
         email_message = EmailMultiAlternatives(subject, body, from_email, [to_email])
         if html_email_template_name is not None:
             html_email = loader.render_to_string(html_email_template_name, context)
             email_message.attach_alternative(html_email, "text/html")
 
-        email_message.send()
+        try:
+            email_message.send()
+        except Exception:
+            logger.exception(
+                "Failed to send password reset email to %s", context["user"].pk
+            )
 
     def get_users(self, email):
         """Given an email, return matching user(s) who should receive a reset.
 
         This allows subclasses to more easily customize the default policies
         that prevent inactive users and users with unusable passwords from
         resetting their password.
         """
         email_field_name = UserModel.get_email_field_name()
         active_users = UserModel._default_manager.filter(
             **{
                 "%s__iexact" % email_field_name: email,
                 "is_active": True,
             }
         )
         return (
             u
             for u in active_users
             if u.has_usable_password()
             and _unicode_ci_compare(email, getattr(u, email_field_name))
         )
 
     def save(
         self,
         domain_override=None,
         subject_template_name="registration/password_reset_subject.txt",
         email_template_name="registration/password_reset_email.html",
         use_https=False,
         token_generator=default_token_generator,
         from_email=None,
         request=None,
         html_email_template_name=None,
         extra_email_context=None,
     ):
         """
         Generate a one-use only link for resetting password and send it to the
         user.
         """
         email = self.cleaned_data["email"]
         if not domain_override:
             current_site = get_current_site(request)
             site_name = current_site.name
             domain = current_site.domain
         else:
             site_name = domain = domain_override
         email_field_name = UserModel.get_email_field_name()
         for user in self.get_users(email):
             user_email = getattr(user, email_field_name)
             context = {
                 "email": user_email,
                 "domain": domain,
                 "site_name": site_name,
                 "uid": urlsafe_base64_encode(force_bytes(user.pk)),
                 "user": user,
                 "token": token_generator.make_token(user),
                 "protocol": "https" if use_https else "http",
                 **(extra_email_context or {}),
             }
             self.send_mail(
                 subject_template_name,
                 email_template_name,
                 context,
                 from_email,
                 user_email,
                 html_email_template_name=html_email_template_name,
             )
diff --git a/docs/ref/logging.txt b/docs/ref/logging.txt
index 8a7e58997e..24ab5d97f3 100644
--- a/docs/ref/logging.txt
+++ b/docs/ref/logging.txt
@@ -209,7 +209,19 @@ Django development server. This logger generates an ``INFO`` message upon
 detecting a modification in a source code file and may produce ``WARNING``
 messages during filesystem inspection and event subscription processes.
 
+.. _django-contrib-auth-logger:
+
+``django.contrib.auth``
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 4.2.16
+
+Log messages related to :doc:`contrib/auth`, particularly ``ERROR`` messages
+are generated when a :class:`~django.contrib.auth.forms.PasswordResetForm` is
+successfully submitted but the password reset email cannot be delivered due to
+a mail sending exception.
+
 .. _django-contrib-gis-logger:
 
 ``django.contrib.gis``
 ~~~~~~~~~~~~~~~~~~~~~~
diff --git a/docs/releases/4.2.16.txt b/docs/releases/4.2.16.txt
index 2a84186867..963036345c 100644
--- a/docs/releases/4.2.16.txt
+++ b/docs/releases/4.2.16.txt
@@ -13,3 +13,14 @@ CVE-2024-45230: Potential denial-of-service vulnerability in ``django.utils.html
 :tfilter:`urlize` and :tfilter:`urlizetrunc` were subject to a potential
 denial-of-service attack via very large inputs with a specific sequence of
 characters.
+
+CVE-2024-45231: Potential user email enumeration via response status on password reset
+======================================================================================
+
+Due to unhandled email sending failures, the
+:class:`~django.contrib.auth.forms.PasswordResetForm` class allowed remote
+attackers to enumerate user emails by issuing password reset requests and
+observing the outcomes.
+
+To mitigate this risk, exceptions occurring during password reset email sending
+are now handled and logged using the :ref:`django-contrib-auth-logger` logger.
diff --git a/docs/releases/5.0.9.txt b/docs/releases/5.0.9.txt
index 50e94ea3f2..52595ae4ff 100644
--- a/docs/releases/5.0.9.txt
+++ b/docs/releases/5.0.9.txt
@@ -13,3 +13,14 @@ CVE-2024-45230: Potential denial-of-service vulnerability in ``django.utils.html
 :tfilter:`urlize` and :tfilter:`urlizetrunc` were subject to a potential
 denial-of-service attack via very large inputs with a specific sequence of
 characters.
+
+CVE-2024-45231: Potential user email enumeration via response status on password reset
+======================================================================================
+
+Due to unhandled email sending failures, the
+:class:`~django.contrib.auth.forms.PasswordResetForm` class allowed remote
+attackers to enumerate user emails by issuing password reset requests and
+observing the outcomes.
+
+To mitigate this risk, exceptions occurring during password reset email sending
+are now handled and logged using the :ref:`django-contrib-auth-logger` logger.
diff --git a/docs/releases/5.1.1.txt b/docs/releases/5.1.1.txt
index dd24929100..8c0468d727 100644
--- a/docs/releases/5.1.1.txt
+++ b/docs/releases/5.1.1.txt
@@ -14,6 +14,17 @@ CVE-2024-45230: Potential denial-of-service vulnerability in ``django.utils.html
 denial-of-service attack via very large inputs with a specific sequence of
 characters.
 
+CVE-2024-45231: Potential user email enumeration via response status on password reset
+======================================================================================
+
+Due to unhandled email sending failures, the
+:class:`~django.contrib.auth.forms.PasswordResetForm` class allowed remote
+attackers to enumerate user emails by issuing password reset requests and
+observing the outcomes.
+
+To mitigate this risk, exceptions occurring during password reset email sending
+are now handled and logged using the :ref:`django-contrib-auth-logger` logger.
+
 Bugfixes
 ========
 
diff --git a/docs/topics/auth/default.txt b/docs/topics/auth/default.txt
index 15af779dfe..4b40cb16f2 100644
--- a/docs/topics/auth/default.txt
+++ b/docs/topics/auth/default.txt
@@ -1644,145 +1644,147 @@ having to write forms for this functionality, the authentication system
 provides several built-in forms located in :mod:`django.contrib.auth.forms`:
 
 .. note::
     The built-in authentication forms make certain assumptions about the user
     model that they are working with. If you're using a :ref:`custom user model
     <auth-custom-user>`, it may be necessary to define your own forms for the
     authentication system. For more information, refer to the documentation
     about :ref:`using the built-in authentication forms with custom user models
     <custom-users-and-the-built-in-auth-forms>`.
 
 .. class:: AdminPasswordChangeForm
 
     A form used in the admin interface to change a user's password, including
     the ability to set an :meth:`unusable password
     <django.contrib.auth.models.User.set_unusable_password>`, which blocks the
     user from logging in with password-based authentication.
 
     Takes the ``user`` as the first positional argument.
 
     .. versionchanged:: 5.1
 
         Option to disable (or reenable) password-based authentication was
         added.
 
 .. class:: AdminUserCreationForm
 
     .. versionadded:: 5.1.1
 
     A form used in the admin interface to create a new user. Inherits from
     :class:`UserCreationForm`.
 
     It includes an additional ``usable_password`` field, enabled by default. If
     ``usable_password`` is enabled, it verifies that ``password1`` and
     ``password2`` are non empty and match, validates the password using
     :func:`~django.contrib.auth.password_validation.validate_password`, and
     sets the user's password using
     :meth:`~django.contrib.auth.models.User.set_password()`.
     If ``usable_password`` is disabled, no password validation is done, and
     password-based authentication is disabled for the user by calling
     :meth:`~django.contrib.auth.models.User.set_unusable_password()`.
 
 .. class:: AuthenticationForm
 
     A form for logging a user in.
 
     Takes ``request`` as its first positional argument, which is stored on the
     form instance for use by sub-classes.
 
     .. method:: confirm_login_allowed(user)
 
         By default, ``AuthenticationForm`` rejects users whose ``is_active``
         flag is set to ``False``. You may override this behavior with a custom
         policy to determine which users can log in. Do this with a custom form
         that subclasses ``AuthenticationForm`` and overrides the
         ``confirm_login_allowed()`` method. This method should raise a
         :exc:`~django.core.exceptions.ValidationError` if the given user may
         not log in.
 
         For example, to allow all users to log in regardless of "active"
         status::
 
             from django.contrib.auth.forms import AuthenticationForm
 
 
             class AuthenticationFormWithInactiveUsersOkay(AuthenticationForm):
                 def confirm_login_allowed(self, user):
                     pass
 
         (In this case, you'll also need to use an authentication backend that
         allows inactive users, such as
         :class:`~django.contrib.auth.backends.AllowAllUsersModelBackend`.)
 
         Or to allow only some active users to log in::
 
             class PickyAuthenticationForm(AuthenticationForm):
                 def confirm_login_allowed(self, user):
                     if not user.is_active:
                         raise ValidationError(
                             _("This account is inactive."),
                             code="inactive",
                         )
                     if user.username.startswith("b"):
                         raise ValidationError(
                             _("Sorry, accounts starting with 'b' aren't welcome here."),
                             code="no_b_users",
                         )
 
 .. class:: BaseUserCreationForm
 
     A :class:`~django.forms.ModelForm` for creating a new user. This is the
     recommended base class if you need to customize the user creation form.
 
     It has three fields: ``username`` (from the user model), ``password1``,
     and ``password2``. It verifies that ``password1`` and ``password2`` match,
     validates the password using
     :func:`~django.contrib.auth.password_validation.validate_password`, and
     sets the user's password using
     :meth:`~django.contrib.auth.models.User.set_password()`.
 
 .. class:: PasswordChangeForm
 
     A form for allowing a user to change their password.
 
 .. class:: PasswordResetForm
 
     A form for generating and emailing a one-time use link to reset a
     user's password.
 
     .. method:: send_mail(subject_template_name, email_template_name, context, from_email, to_email, html_email_template_name=None)
 
         Uses the arguments to send an ``EmailMultiAlternatives``.
-        Can be overridden to customize how the email is sent to the user.
+        Can be overridden to customize how the email is sent to the user. If
+        you choose to override this method, be mindful of handling potential
+        exceptions raised due to email sending failures.
 
         :param subject_template_name: the template for the subject.
         :param email_template_name: the template for the email body.
         :param context: context passed to the ``subject_template``,
             ``email_template``, and ``html_email_template`` (if it is not
             ``None``).
         :param from_email: the sender's email.
         :param to_email: the email of the requester.
         :param html_email_template_name: the template for the HTML body;
             defaults to ``None``, in which case a plain text email is sent.
 
         By default, ``save()`` populates the ``context`` with the
         same variables that
         :class:`~django.contrib.auth.views.PasswordResetView` passes to its
         email context.
 
 .. class:: SetPasswordForm
 
     A form that lets a user change their password without entering the old
     password.
 
 .. class:: UserChangeForm
 
     A form used in the admin interface to change a user's information and
     permissions.
 
 .. class:: UserCreationForm
 
     Inherits from :class:`BaseUserCreationForm`. To help prevent confusion with
     similar usernames, the form doesn't allow usernames that differ only in
     case.
 
 .. currentmodule:: django.contrib.auth
diff --git a/tests/auth_tests/test_forms.py b/tests/auth_tests/test_forms.py
index 13f86b6b77..9463f07d57 100644
--- a/tests/auth_tests/test_forms.py
+++ b/tests/auth_tests/test_forms.py
@@ -1137,231 +1137,252 @@ class UserChangeFormTest(TestDataMixin, TestCase):
 @override_settings(TEMPLATES=AUTH_TEMPLATES)
 class PasswordResetFormTest(TestDataMixin, TestCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
         # This cleanup is necessary because contrib.sites cache
         # makes tests interfere with each other, see #11505
         Site.objects.clear_cache()
 
     def create_dummy_user(self):
         """
         Create a user and return a tuple (user_object, username, email).
         """
         username = "jsmith"
         email = "jsmith@example.com"
         user = User.objects.create_user(username, email, "test123")
         return (user, username, email)
 
     def test_invalid_email(self):
         data = {"email": "not valid"}
         form = PasswordResetForm(data)
         self.assertFalse(form.is_valid())
         self.assertEqual(form["email"].errors, [_("Enter a valid email address.")])
 
     def test_user_email_unicode_collision(self):
         User.objects.create_user("mike123", "mike@example.org", "test123")
         User.objects.create_user("mike456", "mıke@example.org", "test123")
         data = {"email": "mıke@example.org"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(mail.outbox[0].to, ["mıke@example.org"])
 
     def test_user_email_domain_unicode_collision(self):
         User.objects.create_user("mike123", "mike@ixample.org", "test123")
         User.objects.create_user("mike456", "mike@ıxample.org", "test123")
         data = {"email": "mike@ıxample.org"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(mail.outbox[0].to, ["mike@ıxample.org"])
 
     def test_user_email_unicode_collision_nonexistent(self):
         User.objects.create_user("mike123", "mike@example.org", "test123")
         data = {"email": "mıke@example.org"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 0)
 
     def test_user_email_domain_unicode_collision_nonexistent(self):
         User.objects.create_user("mike123", "mike@ixample.org", "test123")
         data = {"email": "mike@ıxample.org"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 0)
 
     def test_nonexistent_email(self):
         """
         Test nonexistent email address. This should not fail because it would
         expose information about registered users.
         """
         data = {"email": "foo@bar.com"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         self.assertEqual(len(mail.outbox), 0)
 
     def test_cleaned_data(self):
         (user, username, email) = self.create_dummy_user()
         data = {"email": email}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         form.save(domain_override="example.com")
         self.assertEqual(form.cleaned_data["email"], email)
         self.assertEqual(len(mail.outbox), 1)
 
     def test_custom_email_subject(self):
         data = {"email": "testclient@example.com"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         # Since we're not providing a request object, we must provide a
         # domain_override to prevent the save operation from failing in the
         # potential case where contrib.sites is not installed. Refs #16412.
         form.save(domain_override="example.com")
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(mail.outbox[0].subject, "Custom password reset on example.com")
 
     def test_custom_email_constructor(self):
         data = {"email": "testclient@example.com"}
 
         class CustomEmailPasswordResetForm(PasswordResetForm):
             def send_mail(
                 self,
                 subject_template_name,
                 email_template_name,
                 context,
                 from_email,
                 to_email,
                 html_email_template_name=None,
             ):
                 EmailMultiAlternatives(
                     "Forgot your password?",
                     "Sorry to hear you forgot your password.",
                     None,
                     [to_email],
                     ["site_monitor@example.com"],
                     headers={"Reply-To": "webmaster@example.com"},
                     alternatives=[
                         ("Really sorry to hear you forgot your password.", "text/html")
                     ],
                 ).send()
 
         form = CustomEmailPasswordResetForm(data)
         self.assertTrue(form.is_valid())
         # Since we're not providing a request object, we must provide a
         # domain_override to prevent the save operation from failing in the
         # potential case where contrib.sites is not installed. Refs #16412.
         form.save(domain_override="example.com")
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(mail.outbox[0].subject, "Forgot your password?")
         self.assertEqual(mail.outbox[0].bcc, ["site_monitor@example.com"])
         self.assertEqual(mail.outbox[0].content_subtype, "plain")
 
     def test_preserve_username_case(self):
         """
         Preserve the case of the user name (before the @ in the email address)
         when creating a user (#5605).
         """
         user = User.objects.create_user("forms_test2", "tesT@EXAMple.com", "test")
         self.assertEqual(user.email, "tesT@example.com")
         user = User.objects.create_user("forms_test3", "tesT", "test")
         self.assertEqual(user.email, "tesT")
 
     def test_inactive_user(self):
         """
         Inactive user cannot receive password reset email.
         """
         (user, username, email) = self.create_dummy_user()
         user.is_active = False
         user.save()
         form = PasswordResetForm({"email": email})
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 0)
 
     def test_unusable_password(self):
         user = User.objects.create_user("testuser", "test@example.com", "test")
         data = {"email": "test@example.com"}
         form = PasswordResetForm(data)
         self.assertTrue(form.is_valid())
         user.set_unusable_password()
         user.save()
         form = PasswordResetForm(data)
         # The form itself is valid, but no email is sent
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 0)
 
     def test_save_plaintext_email(self):
         """
         Test the PasswordResetForm.save() method with no html_email_template_name
         parameter passed in.
         Test to ensure original behavior is unchanged after the parameter was added.
         """
         (user, username, email) = self.create_dummy_user()
         form = PasswordResetForm({"email": email})
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(len(mail.outbox), 1)
         message = mail.outbox[0].message()
         self.assertFalse(message.is_multipart())
         self.assertEqual(message.get_content_type(), "text/plain")
         self.assertEqual(message.get("subject"), "Custom password reset on example.com")
         self.assertEqual(len(mail.outbox[0].alternatives), 0)
         self.assertEqual(message.get_all("to"), [email])
         self.assertTrue(
             re.match(r"^http://example.com/reset/[\w+/-]", message.get_payload())
         )
 
     def test_save_html_email_template_name(self):
         """
         Test the PasswordResetForm.save() method with html_email_template_name
         parameter specified.
         Test to ensure that a multipart email is sent with both text/plain
         and text/html parts.
         """
         (user, username, email) = self.create_dummy_user()
         form = PasswordResetForm({"email": email})
         self.assertTrue(form.is_valid())
         form.save(
             html_email_template_name="registration/html_password_reset_email.html"
         )
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(len(mail.outbox[0].alternatives), 1)
         message = mail.outbox[0].message()
         self.assertEqual(message.get("subject"), "Custom password reset on example.com")
         self.assertEqual(len(message.get_payload()), 2)
         self.assertTrue(message.is_multipart())
         self.assertEqual(message.get_payload(0).get_content_type(), "text/plain")
         self.assertEqual(message.get_payload(1).get_content_type(), "text/html")
         self.assertEqual(message.get_all("to"), [email])
         self.assertTrue(
             re.match(
                 r"^http://example.com/reset/[\w/-]+",
                 message.get_payload(0).get_payload(),
             )
         )
         self.assertTrue(
             re.match(
                 r'^<html><a href="http://example.com/reset/[\w/-]+/">Link</a></html>$',
                 message.get_payload(1).get_payload(),
             )
         )
 
+    @override_settings(EMAIL_BACKEND="mail.custombackend.FailingEmailBackend")
+    def test_save_send_email_exceptions_are_catched_and_logged(self):
+        (user, username, email) = self.create_dummy_user()
+        form = PasswordResetForm({"email": email})
+        self.assertTrue(form.is_valid())
+
+        with self.assertLogs("django.contrib.auth", level=0) as cm:
+            form.save()
+
+        self.assertEqual(len(mail.outbox), 0)
+        self.assertEqual(len(cm.output), 1)
+        errors = cm.output[0].split("\n")
+        pk = user.pk
+        self.assertEqual(
+            errors[0],
+            f"ERROR:django.contrib.auth:Failed to send password reset email to {pk}",
+        )
+        self.assertEqual(
+            errors[-1], "ValueError: FailingEmailBackend is doomed to fail."
+        )
+
     @override_settings(AUTH_USER_MODEL="auth_tests.CustomEmailField")
     def test_custom_email_field(self):
         email = "test@mail.com"
         CustomEmailField.objects.create_user("test name", "test password", email)
         form = PasswordResetForm({"email": email})
         self.assertTrue(form.is_valid())
         form.save()
         self.assertEqual(form.cleaned_data["email"], email)
         self.assertEqual(len(mail.outbox), 1)
         self.assertEqual(mail.outbox[0].to, [email])
 
     def test_html_autocomplete_attributes(self):
         form = PasswordResetForm()
         self.assertEqual(form.fields["email"].widget.attrs["autocomplete"], "email")
diff --git a/tests/mail/custombackend.py b/tests/mail/custombackend.py
index 14e7f077ba..c63f1c07b5 100644
--- a/tests/mail/custombackend.py
+++ b/tests/mail/custombackend.py
@@ -12,3 +12,9 @@ class EmailBackend(BaseEmailBackend):
         # Messages are stored in an instance variable for testing.
         self.test_outbox.extend(email_messages)
         return len(email_messages)
+
+
+class FailingEmailBackend(BaseEmailBackend):
+
+    def send_messages(self, email_messages):
+        raise ValueError("FailingEmailBackend is doomed to fail.")
